### PR TITLE
Review dynamic vector API

### PR DESCRIPTION
**Pull request by lionel-**
As discussed @DavisVaughan 

- Use `dyn_` instead of `arr_`
- Add poker and getter helpers

Also add `r_as_char()`, `r_as_complex()`, and `r_as_double()` helpers for input validation.

### CODE REVIEW

File: src/rlang/vec.c [hunk 1]
```diff
@@ -226,3 +226,41 @@ void r_vec_poke_range(r_obj* x, r_ssize offset,
                       r_obj* y, r_ssize from, r_ssize to) {
   r_vec_poke_n(x, offset, y, from, to - from + 1);
 }
+
+bool _r_is_finite(r_obj* x) {
```
Comment (DavisVaughan) [inline src/rlang/vec.c:230]: Is `_r` meaning that its extremely internal?

File: src/rlang/vec.h [hunk 2]
```diff
@@ -266,6 +278,34 @@ r_obj* r_shared_lgl(bool x) {
   }
 }
 
+static inline
+bool _r_has_correct_length(r_obj* x, r_ssize n) {
+  return n < 0 || r_length(x) == n;
+}
+extern
```
Comment (DavisVaughan) [inline src/rlang/vec.h:285]: Why `extern`?

**Review by DavisVaughan — COMMENTED** [review]
These look great, thanks again!

File: src/rlang/vec.c (hunk 1 above)
Comment (lionel-) [reply src/rlang/vec.c:230]: yes it means it's not part of the API. I had to pull them from `internal/`, but the interface of these functions still need a bit of review.

File: src/rlang/vec.h (hunk 2 above)
Comment (lionel-) [reply src/rlang/vec.h:285]: oh it's not necessary here since it's not a variable. Not sure why I wrote this.

**Pull request merged**